app/src/org/commcare/services/CommCareFirebaseMessagingService.java

```diff
@@ -28,81 +29,95 @@
-        if (payloadData.size() == 0){
+        if (payloadData.isEmpty()) {
             return;
         }
 
+        showNotification(payloadData);
+
         FCMMessageData fcmMessageData = new FCMMessageData(payloadData);
 
-        switch(fcmMessageData.getAction()){
+        switch (fcmMessageData.getAction()) {
             case SYNC -> dataSyncer.syncData(fcmMessageData);
-            default ->
-                    Logger.log(LogTypes.TYPE_FCM, "Invalid FCM action");
+            default -> Logger.log(LogTypes.TYPE_FCM, "Invalid FCM action");
         }
     }
 
     @Override
     public void onNewToken(String token) {
         // TODO: Remove the token from the log
-        Logger.log(LogTypes.TYPE_FCM, "New registration token was generated: "+token);
+        Logger.log(LogTypes.TYPE_FCM, "New registration token was generated: " + token);
         FirebaseMessagingUtil.updateFCMToken(token);
     }
 
 
-   /**
-    * This method purpose is to show notifications to the user when the app is in the foreground.
-    * When the app is in the background, FCM is responsible for notifying the user
-    *
-    */
-    private void showNotification(RemoteMessage.Notification notification) {
-        String notificationTitle = notification.getTitle();
-        String notificationText = notification.getBody();
+    /**
+     * This method purpose is to show notifications to the user when the app is in the foreground.
+     * When the app is in the background, FCM is responsible for notifying the user
+     */
+    private void showNotification(Map<String, String> payloadData) {
+        String notificationTitle = payloadData.get("title");
+        String notificationText = payloadData.get("body");
         NotificationManager mNM = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
 
-        Intent i = new Intent(this, DispatchActivity.class);
-        i.setAction(Intent.ACTION_MAIN);
-        i.addCategory(Intent.CATEGORY_LAUNCHER);
+        Intent intent;
+
+        if (payloadData.containsKey("action")) {
+            intent = new Intent(getApplicationContext(), ConnectActivity.class);
+            intent.putExtra("action", payloadData.get("action"));
+            intent.putExtra("opportunity_id", payloadData.get("opportunity_id"));
+        } else {
+            intent = new Intent(this, DispatchActivity.class);
+            intent.setAction(Intent.ACTION_MAIN);
+            intent.addCategory(Intent.CATEGORY_LAUNCHER);
+        }
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
 
         PendingIntent contentIntent;
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
-            contentIntent = PendingIntent.getActivity(this, 0, i, PendingIntent.FLAG_IMMUTABLE);
-        else
-            contentIntent = PendingIntent.getActivity(this, 0, i, 0);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
```

Comment:
Nit: Code in the if-else can just specify an (int) flags variable, then in one line below:
PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent,  flags);

Reply:
Removed else if condition. looks like that execute same code for both else if and else

Reply:
Hmm, that doesn't look right. The else-if passed FLAG_IMMUTABLE while the else passed 0, do those result in the same behavior?

But my comment was more about not repeating the code that calls getActivity. Since the only thing that changes in the various if-else cases is the flags, let's just choose the flags in the if-structure, then call PendingIntent.getActivity afterward

Reply:
is this looks correct ?

`int flags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
                ? PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_ONE_SHOT
                : PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_ONE_SHOT;

        PendingIntent contentIntent = PendingIntent.getActivity(this, 0, intent, flags);`